tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,753 @@
+	)
+
+	cdm.wbufLock.RLock()
+	if sgmIndex == cdm.curFileSequence && chkStart > int(cdm.n)-cdm.wbuf.Buffered() {
+		chunk := cdm.getChunkFromBuffer(ref)
+		if chunk != nil {
+			cdm.wbufLock.RUnlock()
+			return chunk, nil
+		}
+	}
+	cdm.wbufLock.RUnlock()
+
+	cdm.mmappedChunksMtx.RLock()
+	// We hold this read lock for the entire duration because if the Close()
+	// is called, the data in the byte slice will get corrupted as the mmapped
+	// file will be closed.
+	defer cdm.mmappedChunksMtx.RUnlock()
+
+	if cdm.closed {
+		return nil, ErrChunkDiskMapperClosed
+	}
+
+	sgmBytes, ok := cdm.mmappedChunkFiles[sgmIndex]
+	if !ok {
+		if sgmIndex > cdm.curFileSequence {
+			return nil, errors.Errorf("head chunk file index %d more than current open file", sgmIndex)
+		}
+		return nil, errors.Errorf("head chunk file index %d does not exist on disk", sgmIndex)
+	}
+
+	if chkStart+MaxChunkLengthFieldSize > sgmBytes.Len() {
+		return nil, errors.Errorf("head chunk file doesn't include enough bytes to read the chunk size data field - required:%v, available:%v, file:%d", chkStart+MaxChunkLengthFieldSize, sgmBytes.Len(), sgmIndex)
+	}
+
+	// Encoding.
+	chkEnc := sgmBytes.Range(chkStart, chkStart+ChunkEncodingSize)[0]
+
+	// Data length.
+	// With the minimum chunk length this should never cause us reading
+	// over the end of the slice.
+	chkDataLenStart := chkStart + ChunkEncodingSize
+	c := sgmBytes.Range(chkDataLenStart, chkDataLenStart+MaxChunkLengthFieldSize)
+	chkDataLen, n := binary.Uvarint(c)
+	if n <= 0 {
+		return nil, errors.Errorf("reading chunk length failed with %d", n)
+	}
+
+	// Verify the chunk data end.
+	chkDataEnd := chkDataLenStart + n + int(chkDataLen)
+	if chkDataEnd > sgmBytes.Len() {
+		return nil, errors.Errorf("head chunk file doesn't include enough bytes to read the chunk - required:%v, available:%v", chkDataEnd, sgmBytes.Len())
+	}
+
+	// Check the CRC.
+	sum := sgmBytes.Range(chkDataEnd, chkDataEnd+CRCSize)
+	if _, err := chkCRC32.Write(sgmBytes.Range(chkStart-(SeriesRefSize+2*MintMaxtSize), chkDataEnd)); err != nil {
+		return nil, err
+	}
+	if act := chkCRC32.Sum(nil); !bytes.Equal(act, sum) {
+		return nil, errors.Errorf("checksum mismatch expected:%x, actual:%x", sum, act)
+	}
+
+	// The chunk data itself.
+	chkData := sgmBytes.Range(chkDataEnd-int(chkDataLen), chkDataEnd)
+	return cdm.pool.Get(chunkenc.Encoding(chkEnc), chkData)
+}
+
+func (cdm *ChunkDiskMapper) getChunkFromBuffer(chunkRef uint64) chunkenc.Chunk {
+	shardIdx := chunkRef % inBufferShards
+
+	cdm.inBufferChunksLocks[shardIdx].RLock()
+	defer cdm.inBufferChunksLocks[shardIdx].RUnlock()
+
+	return cdm.inBufferChunks[shardIdx][chunkRef]
+}
+
+// IterateAllChunks iterates on all the chunks in its byte slices in the order of the head chunk file sequence
+// and runs the provided function on each chunk. It returns on the first error encountered.
+func (cdm *ChunkDiskMapper) IterateAllChunks(f func(seriesRef, chunkRef uint64, mint, maxt int64) error) error {
```

Comment:
Is this really needed? It is used only by tests? I guess, yes for DB cut? If no let's move it to test maybe? Or unexport?

Reply:
It will be used by Head when we will replay chunks from the disk. It is already integrated in the parent PR, but there is no use of this as a standalone function in this PR.